srtcore/queue.cpp

```diff
@@ -1098,7 +1098,15 @@ void CRendezvousQueue::updateConnStatus(EReadStatus rst, EConnectStatus cst, con
     {
         HLOGC(cnlog.Debug, log << "updateConnStatus: COMPLETING dep objects update on failed @" << i->id);
         i->u->m_bConnecting = false;
-        i->u->updateBrokenConnection();
+
+        // DO NOT close the socket here because in this case it might be
+        // unable to get status from at the right moment. Also only member
+        // sockets should be taken care of internally - single sockets should
+        // be normally closed by the application, after it is done with them.
+
+        // app can call any UDT API to learn the connection_broken error
+        CUDT::s_UDTUnited.m_EPoll.update_events(i->u->m_SocketID, i->u->m_sPollID, SRT_EPOLL_IN | SRT_EPOLL_OUT | SRT_EPOLL_ERR, true);
```

Comment:
> ... a socket that failed to connect remains in the CONNECTING state and the only sign of error is in epoll ERR flag.

In the current version and in this PR it does not remain in the connecting state.
Does it change the behavior towards 1.4.1 then?

```c++
i->u->m_bConnecting = false;
```

Reply:
This is the code that existed in 1.4.1. This had to be removed in order to move the closing activities outside the lock.
```
        if (CTimer::getTime() >= i->m_ullTTL)
        {
            HLOGC(mglog.Debug,
                  log << "RendezvousQueue: EXPIRED (" << (i->m_ullTTL ? "enforced on FAILURE" : "passed TTL")
                      << ". removing from queue");
            // connection timer expired, acknowledge app via epoll
            i->m_pUDT->m_bConnecting = false;
            CUDT::s_UDTUnited.m_EPoll.update_events(i->m_iID, i->m_pUDT->m_sPollID, UDT_EPOLL_ERR, true);
            /*
             * Setting m_bConnecting to false but keeping socket in rendezvous queue is not a good idea.
             * Next CUDT::close will not remove it from rendezvous queue (because !m_bConnecting)
             * and may crash here on next pass.
             */
            if (AF_INET == i->m_iIPversion)
                delete (sockaddr_in *)i->m_pPeerAddr;
            else
                delete (sockaddr_in6 *)i->m_pPeerAddr;

            // i_next was preincremented, but this is guaranteed to point to
            // the element next to erased one.
            i_next = m_lRendezvousID.erase(i);
            continue;
        }

```

Reply:
Then it does not remain in the CONNECTING state, does it?

Reply:
Changing these flags have nothing to do with the "socket state". The socket state for a connection-timedout socket is still `SRTS_CONNECTING`.

Reply:
So internally it will have `m_Status == SRTS_CONNECTING`, but the status from the API seems to be "BROKEN" when not connecting and not connected.
A bit confusing, but the motivation seems to be to have a socket in a BROKEN state, but not scheduled for closing by GC.

```c++
SRT_SOCKSTATUS CUDTSocket::getStatus()
{
    // TTL in CRendezvousQueue::updateConnStatus() will set m_bConnecting to false.
    // Although m_Status is still SRTS_CONNECTING, the connection is in fact to be closed due to TTL expiry.
    // In this case m_bConnected is also false. Both checks are required to avoid hitting
    // a regular state transition from CONNECTING to CONNECTED.

    if (m_pUDT->m_bBroken)
        return SRTS_BROKEN;

    // Connecting timed out
    if ((m_Status == SRTS_CONNECTING) && !m_pUDT->m_bConnecting && !m_pUDT->m_bConnected)
        return SRTS_BROKEN;

    return m_Status;
}
```